crates/ide_assists/src/handlers/early_return.rs

```diff
@@ -143,10 +143,25 @@ pub(crate) fn convert_to_guarded_return(acc: &mut Assists, ctx: &AssistContext)
                         make::expr_match(cond_expr, make::match_arm_list(vec![happy_arm, sad_arm]))
                     };
 
-                    let let_stmt = make::let_stmt(
-                        make::ident_pat(make::name(&bound_ident.syntax().to_string())).into(),
-                        Some(match_expr),
-                    );
+                    let pattern = if bound_ident
+                        .syntax()
+                        .children_with_tokens()
+                        .filter_map(|it| it.into_token())
+                        .find(|it| it.kind() == T![mut])
+                        .is_some()
+                    {
+                        make::ident_mut_pat(make::name(
+                            &bound_ident.syntax().last_token().unwrap().to_string(),
+                        ))
+                        .into()
+                    } else {
+                        make::ident_pat(make::name(
+                            &bound_ident.syntax().last_token().unwrap().to_string(),
+                        ))
+                        .into()
+                    };
+
+                    let let_stmt = make::let_stmt(pattern, Some(match_expr));
```

Comment:
Mmh, this doesn't handle `ref` patterns either and will crash then still. Ideally we should filter the pattern out some more earlier.

I think we want to check if the pattern here, the one in `bound_ident` is an `IdentPat` and store that directly there instead of the general `Pat`.
https://github.com/rust-analyzer/rust-analyzer/blob/065a8e87cd015ccfe5c01ad6b603f4fdad18f768/crates/ide_assists/src/handlers/early_return.rs#L57-L60
That should make stuff a lot simpler here as we can just re-use the bound_ident in `make::let_stmt` there while keeping the possible `mut` and `ref` tokens. Should also remove the need of reconstructing a pattern all together there.

Reply:
How about now?